File: d2/d2.hs
#! /usr/bin/env nix-shell
#! nix-shell -p "haskellPackages.ghcWithPackages (pkgs: with pkgs; [])" -i "runghc -Wall"

{-# LANGUAGE LambdaCase #-}

import Text.Read (readEither)

data Line = Line Int Int Char String

readLine :: String -> Either String Line
readLine s =
  let [range, [seek, _], input] = words s
      -- There's no splitOn in the Prelude
      [lower, upper] = words [if c == '-' then ' ' else c | c <- range]
   in do
        l <- readEither lower
        u <- readEither upper
        Right $ Line l u seek input

isValid :: Line -> Bool
isValid (Line lower upper seek input) =
  let count = length $ filter ((==) seek) input
   in count >= lower && count <= upper

isValidPartTwo :: Line -> Bool
isValidPartTwo (Line lower upper seek input) =
  let pos1 = input !! (lower - 1)
      pos2 = input !! (upper - 1)
   in case (pos1 == seek, pos2 == seek) of
        (True, False) -> True
        (False, True) -> True
        _ -> False

main :: IO ()
main =
  sequence . map readLine . lines <$> getContents >>= \case
    Left e -> print e
    Right success -> do
      print $ "Valid passwords part one: " <> show (length $ filter isValid success)
      print $ "Valid passwords part two: " <> show (length $ filter isValidPartTwo success)

File: d2/d2.lua
#!/usr/bin/env nix-shell
--[[
#! nix-shell -p "lua53Packages.lua" -i lua
]]

local s = io.read("*a")
local valid = 0
local valid_part_two = 0
-- Iterate linewise
for line in s:gmatch("[^\r\n]+") do
	local lower, upper, seek, input = line:match("(%d+)-(%d+) (%l): (%l+)")
	lower = tonumber(lower)
	upper = tonumber(upper)

	local count = 0
	for _ in string.gmatch(input, string.format("%s", seek)) do
		count = count + 1
	end

	if count >= lower and count <= upper then
		valid = valid + 1
	end

	-- No string indexing in lua, get substring from START to FINISH with
	-- both set to same index
	local pos1 = input:sub(lower, lower) == seek
	local pos2 = input:sub(upper, upper) == seek

	if (pos1 and not pos2) or (pos2 and not pos1) then
		valid_part_two = valid_part_two + 1
	end
end
print("Part One: ", valid)
print("Part Two: ", valid_part_two)
